File: Pokemon/frameworks/runtime-src/Classes/Framework/net/NetCenter.cpp
#include "NetCenter.h"
#include "NetRequest.h"
#include "NetResponse.h"
#include "NetService.h"
#include "NetHandler.h"
#include "NetFilter.h"
#include "NetProtocol.h"

using namespace std;
using namespace cocos2d;

namespace framework
{
	NetCenter::NetCenter()
		: _services(new Map())
		, _protocols(new Map())
		, _filters(new Vector())
	{
	}

	NetCenter::~NetCenter()
	{
		CC_SAFE_DELETE(_services);
		CC_SAFE_DELETE(_protocols);
		CC_SAFE_DELETE(_filters);
	}

	void NetCenter::sendMessage(framework::NetRequest *request)
	{
		string protocolId = request->getProtocolId();
		NetProtocol *pProtocol = (NetProtocol*)_protocols->objectForKey(protocolId);
		CCASSERT(pProtocol, "unknown protocol");
		NetService *pService = (NetService*)_services->objectForKey(pProtocol->getServiceId());
		CCASSERT(pService, "unknown service");
		pService->sendMessage(request);
	}

	void NetCenter::sendCommand(const std::string &serviceId, const std::string &cmd, cocos2d::Ref *params)
	{
		NetService *pService = (NetService*)_services->objectForKey(serviceId);
		if (pService)
		{
			pService->executeCommand(cmd, params);
		}
		else
		{
			CCLOG("The service [%s] doesn't exist.", serviceId.c_str());
		}
	}

	void NetCenter::dispatchSuccessfulMessage(framework::NetResponse *response)
	{
		CCASSERT(response, "Invalid response to dispatch");
		string protocolId = response->getProtocolId();
		NetProtocol *pProtocol = (NetProtocol*)_protocols->objectForKey(protocolId);
		if (!pProtocol)
		{
			CCLOG("Unknown protocol");
			return;
		}
		// filter first
		NetFilter *pFilter = nullptr;
		for (int i = 0; i < _filters->getLength(); ++i)
		{
			pFilter = (NetFilter*)_filters->objectAt(i);
			response = pFilter->filterSuccessfulMessage(response);
		}
		// handle the response
		NetHandler *pHandler = (NetHandler*)pProtocol->getRelatedHandler();
		CCASSERT(pHandler, "The protocol must have a handler to handle the response.");
		pHandler->handleSuccessfulMessage(response);
	}

	void NetCenter::dispatchFailedMessage(framework::NetResponse *response)
	{
		CCASSERT(response, "Invalid response to dispatch");
		string protocolId = response->getProtocolId();
		NetProtocol *pProtocol = (NetProtocol*)_protocols->objectForKey(protocolId);
		if (!pProtocol)
		{
			CCLOG("Unknown protocol");
			return;
		}
		// filter first
		NetFilter *pFilter = nullptr;
		for (int i = 0; i < _filters->getLength(); ++i)
		{
			pFilter = (NetFilter*)_filters->objectAt(i);
			response = pFilter->filterFailedMessage(response);
		}
		// handle the response
		NetHandler *pHandler = (NetHandler*)pProtocol->getRelatedHandler();
		CCASSERT(pHandler, "The protocol must have a handler to handle the response.");
		pHandler->handleFailedMessage(response);
	}

	void NetCenter::addNetService(const std::string &serviceId, framework::NetService *service)
	{
		_services->setObjectForKey(service, serviceId);
	}

	void NetCenter::removeNetService(const std::string &serviceId)
	{
		_services->removeObjectForKey(serviceId);
	}

	void NetCenter::addNetProtocol(const std::string &protocolId, const std::string &serviceId, framework::NetHandler *handler)
	{
		auto pProtocol = NetProtocol::create(protocolId, serviceId, handler);
		this->addNetProtocol(pProtocol);
	}

	void NetCenter::addNetProtocol(framework::NetProtocol *protocol)
	{
		_protocols->setObjectForKey(protocol, protocol->getProtocolId());
	}

	void NetCenter::removeNetProtocol(const std::string &protocolId)
	{
		_protocols->removeObjectForKey(protocolId);
	}

	void NetCenter::addNetFilter(NetFilter *filter)
	{
		_filters->addObject(filter);
	}

	void NetCenter::removeNetFilter(NetFilter *filter)
	{
		_filters->removeObject(filter);
	}

	void NetCenter::removeAllNetFilters()
	{
		_filters->clear();
	}
}

File: Pokemon/frameworks/psframework/scripting/lua-bindings/auto/api/NetCenter.lua

--------------------------------
-- @module NetCenter
-- @extend Ref

--------------------------------
-- @function [parent=#NetCenter] addNetService 
-- @param self
-- @param #string str
-- @param #pf.NetService netservice
        
--------------------------------
-- @function [parent=#NetCenter] removeAllNetFilters 
-- @param self
        
--------------------------------
-- @function [parent=#NetCenter] addNetFilter 
-- @param self
-- @param #pf.NetFilter netfilter
        
--------------------------------
-- @function [parent=#NetCenter] removeNetProtocol 
-- @param self
-- @param #string str
        
--------------------------------
-- @function [parent=#NetCenter] removeNetFilter 
-- @param self
-- @param #pf.NetFilter netfilter
        
--------------------------------
-- @function [parent=#NetCenter] removeNetService 
-- @param self
-- @param #string str
        
--------------------------------
-- @function [parent=#NetCenter] sendMessage 
-- @param self
-- @param #pf.NetRequest netrequest
        
--------------------------------
-- overload function: addNetProtocol(pf.NetProtocol)
--          
-- overload function: addNetProtocol(string, string, pf.NetHandler)
--          
-- @function [parent=#NetCenter] addNetProtocol
-- @param self
-- @param #string str
-- @param #string str
-- @param #pf.NetHandler nethandler

--------------------------------
-- @function [parent=#NetCenter] dispatchFailedMessage 
-- @param self
-- @param #pf.NetResponse netresponse
        
--------------------------------
-- @function [parent=#NetCenter] dispatchSuccessfulMessage 
-- @param self
-- @param #pf.NetResponse netresponse
        
--------------------------------
-- @function [parent=#NetCenter] sendCommand 
-- @param self
-- @param #string str
-- @param #string str
-- @param #cc.Ref ref
        
--------------------------------
-- @function [parent=#NetCenter] getInstance 
-- @param self
-- @return NetCenter#NetCenter ret (return value: pf.NetCenter)
        
return nil
